src/act.other.cpp

```diff
@@ -4165,6 +4165,128 @@ ACMD(do_memory)
+{
+  struct obj_data *deck;
+  if (!ch) {
+    mudlog("SYSERR: Got NULL ch to find_cyberdeck()!", ch, LOG_SYSLOG, TRUE);
+    return NULL;
+  }
+
+  for (deck = ch->carrying; deck; deck = deck->next_content) {
+    if (OBJ_IS_VALID_CYBERDECK(deck)) {
+      return deck;
+    }
+  }
+
+  for (int wear_idx = 0; !deck && wear_idx < NUM_WEARS; wear_idx++) {
+    if (OBJ_IS_VALID_CYBERDECK(GET_EQ(ch, wear_idx))) {
+      return GET_EQ(ch, wear_idx);
+    }
+  }
+
+  return NULL;
+}
+
+ACMD(do_download_headware)
+{
+  struct obj_data *deck, *jack, *memory;
+  skip_spaces(&argument);
+  if (!*argument) {
+    send_to_char("Which file number do you want to transfer out of headware memory?\r\n", ch);
+    return;
+  }
+
+  for (memory = ch->cyberware; memory; memory = memory->next_content) {
+    if (GET_OBJ_VAL(memory, 0) == CYB_MEMORY) {
+      break;
+    }
+  }
+  if (!memory) {
+    send_to_char("You have no headware memory.\r\n", ch);
+    return;
+  }
+
+  // Error message sent in function.
+  if (!(jack = get_datajack(ch, FALSE))) {
+    return;
+  }
+
+  if (!(deck = find_cyberdeck(ch))) {
+    send_to_char(ch, "You need to have a working cyberdeck to transfer data out of headware memory.\r\n");
+    return;
+  }
+
+  int x = atoi(argument);
+  struct obj_data *obj = memory->contains;
+  for (; obj; obj = obj->next_content) {
+    if (!--x)
+      break;
+  }
+  if (!obj) {
+    send_to_char("You don't have that many files in your memory.\r\n", ch);
+  } else if (GET_OBJ_TYPE(obj) != ITEM_DECK_ACCESSORY) {
+    send_to_char("You can only download photos and data.\r\n", ch);
+  } else {
+    obj_from_obj(obj);
+    GET_CYBERWARE_MEMORY_USED(memory) -= GET_CHIP_SIZE(obj) - GET_CHIP_COMPRESSION_FACTOR(obj);
+    obj_to_char(obj, ch);
+    send_to_char(ch, "You transfer %s through %s into your hands.\r\n", GET_OBJ_NAME(obj), GET_OBJ_NAME(deck));
+  }
+}
+
+ACMD(do_upload_headware)
+{
+  struct obj_data *deck, *jack, *memory, *obj;
+
+  skip_spaces(&argument);
+  if (!*argument) {
+    send_to_char("Which file do you want to transfer into headware memory?\r\n", ch);
+    return;
+  }
+
+  for (memory = ch->cyberware; memory; memory = memory->next_content) {
+    if (GET_OBJ_VAL(memory, 0) == CYB_MEMORY) {
+      break;
+    }
+  }
+  if (!memory) {
+    send_to_char("You have no headware memory.\r\n", ch);
+    return;
+  }
+
+  // Error message sent in function.
+  if (!(jack = get_datajack(ch, FALSE))) {
+    return;
+  }
+
+  if (!(deck = find_cyberdeck(ch))) {
+    send_to_char(ch, "You need to have a working cyberdeck to transfer data into headware memory.\r\n");
+    return;
+  }
+
+  obj = get_obj_in_list_vis(ch, argument, ch->carrying);
+  if (!obj) {
+    send_to_char(ch, "You aren't carrying that file.\r\n");
+    return;
+  } else if (GET_OBJ_TYPE(obj) != ITEM_DECK_ACCESSORY) {
+    send_to_char(ch, "You can only upload photos and data.\r\n");
+    return;
+  }
+
+  if (GET_CHIP_SIZE(obj) > GET_CYBERWARE_MEMORY_FREE(memory)) {
```

Comment:
It's been a while since I looked at the headware memory code, so I don't recall the specifics of how compression works, but you checked the compression factor when uninstalling it and don't check it here. Is this intended? Does that clear when it's uninstalled somehow?

Reply:
I'll have to double check this.  I assumed that compression is only possible in place, that is, that it reduces the used memory after the full amount has already occupied the space.  So when removing the file you need to adjust the used memory based on compression, but when adding it, the compression is not considered

Reply:
`help compress` says `Files that are compressed must be decompressed before they are used.` so I'm going to remove the compression math and just ensure that compressed files can't be downloaded